src/audio/google/google_rtc_audio_processing.c

```diff
@@ -83,30 +110,131 @@ void GoogleRtcFree(void *ptr)
 	return rfree(ptr);
 }
 
-#if CONFIG_IPC_MAJOR_4
-static void google_rtc_audio_processing_params(struct processing_module *mod)
+static ALWAYS_INLINE float s16_to_float(const char *ptr)
 {
-	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
-	struct sof_ipc_stream_params *params = mod->stream_params;
-	struct comp_buffer *sinkb, *sourceb;
-	struct list_item *source_list;
-	struct comp_dev *dev = mod->dev;
+	float scale = -(float)SHRT_MIN;
+	float x = *(int16_t *)ptr;
 
-	ipc4_base_module_cfg_to_stream_params(&mod->priv.cfg.base_cfg, params);
-	component_set_nearest_period_frames(dev, params->rate);
+	return (1.0f / scale) * x;
+}
+
+static ALWAYS_INLINE void float_to_s16(float x, char *dst)
```

Comment:
wouldn't making the second argument an `int16_t *` pointer help the compiler detect potential alignment issues? Similar other cases below too

Reply:
We need a single type for the function pointer being passed through the middle layers.  Using the "correct" type would force a cast (and thus wouldn't provide any safety anyway).  Again, the only reason this looks like this is to get the code generation to parametrize the inner loop correctly, so I can have one function that does "copy to/from the stream" and another that generates it template-style in different ways for different formats. 
 Otherwise there would need to be eight (!) essentially verbatim cut-and-paste copies of mostly the same code.

At runtime, the ALWAYS_INLINE cleans everything up.  This function never exists as a separate entity at runtime and that address is never taken as a true pointer.